### PR TITLE
[uv] port pyright script from pip -> uv

### PR DESCRIPTION
**Pull request by smackesey**
## Summary & Motivation

Companion PR: https://github.com/dagster-io/internal/pull/8518

Make `scripts/run-pyright.py` use `uv` instead of `pip`. Experiment with size n=2 on BK shows a ~5-10m speedup in the OSS pyright step from 14m40s to (9m39s, 6m13s).

`uv` also integrates some utilities from `pip-tools` that may be able to improve `make rebuild_pyright_pins` (though the process is already sped up a lot by this PR). I'll attempt to incorporate this functionality in a future PR.

## Notes on editable installs

For some reason, even though no editable installs are altered in the pins, two imports in our examples started being flagged as type errors-- in both cases it was a sibling module import being flagged as unresolvable. Must be some subtle difference in how editable installs are being done between `pip` and `uv`, even with `--editable-mode=compat` set. In any case I was able to resolve the errors by making these two instances relative imports, which is arguably better in these cases anyway as it doesn't implicitly assume a working directory.

## Notes on `uv venv` python interpreter resolution

The most tricky part of this was discovering and correcting for the python resolution logic `uv` uses to create virtual environments. Unlike the built-in `venv` package, `uv` (because it does not actually use python at all) _does not use the python it is installed under by default_. That is:

```
$ which pip  # /some/pip (which goes with /some/python)
$ pip install uv
$ uv venv .venv  # the created venv may not be using /some/python!
```

This had significant consequences on BK internal. The images used in internal contain a python at both:

- `/bin/python`
- `/usr/local/bin/python`

The one at `/usr/local/bin/python` is the one used for all our tests and is the default target for `pip install`. But when creating a virtual env, `uv` actually chooses `/bin/python` by default. At time of this PR, the pyright env cannot be built against `/bin/python` (3.11.2) for obscure reasons having to do with PEP 517 build isolation and the failure of `pendulum==2.1.2` to accurately require `setuptools` as a buildtime dependency 💀. There's also some other nonsense with `xmlsec`. To fix these issues I had to point `uv` to `/usr/local/bin/python` when creating the venv on BK internal. 

## How I Tested These Changes

BK

### CODE REVIEW

**Comment by smackesey** [comment]
* **#20199** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/20199?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @smackesey and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: scripts/run-pyright.py [hunk 1]
```diff
@@ -313,18 +333,42 @@ def normalize_env(env: str, rebuild: bool, update_pins: bool) -> None:
...
-    oss_root = "${DAGSTER_GIT_REPO_DIR}/" if is_internal else ""
-    dep_list = re.sub(
-        r"-e git.*?dagster-io/dagster.*?\&subdirectory=(.+)", f"-e {oss_root}\\1", raw_dep_list
-    )
-    dep_list = re.sub(r"-e git.*?dagster-io/internal.*?\&subdirectory=(.+)", "-e \\1", dep_list)
+    oss_root_ref = "${DAGSTER_GIT_REPO_DIR}/" if is_internal else ""
+    resolved_oss_root = os.environ["DAGSTER_GIT_REPO_DIR"].rstrip("/")
+    resolved_internal_root = os.environ["DAGSTER_INTERNAL_GIT_REPO_DIR"].rstrip("/")
+    dep_list = re.sub(f"-e file://{resolved_oss_root}/(.+)", f"-e {oss_root_ref}\\1", raw_dep_list)
+    dep_list = re.sub(f"-e file://{resolved_internal_root}/(.+)", "-e \\1", dep_list)
```
Comment (smackesey) [inline scripts/run-pyright.py:371]: The output format of `uv pip freeze` is slightly different than `pip freeze` so has to be differently massaged.

File: .buildkite/dagster-buildkite/dagster_buildkite/steps/dagster.py [hunk 2]
```diff
@@ -107,7 +107,7 @@ def build_repo_wide_pyright_steps() -> List[CommandStep]:
         CommandStepBuilder(":pyright: pyright")
         .run(
             "curl https://sh.rustup.rs -sSf | sh -s -- --default-toolchain nightly -y",
-            "pip install -e python_modules/dagster[pyright] -e python_modules/dagster-pipes",
+            "pip install uv -e python_modules/dagster[pyright] -e python_modules/dagster-pipes",
```
[the remark this answers is not in the document]
Comment (smackesey) [reply .buildkite/dagster-buildkite/dagster_buildkite/steps/dagster.py:110]: Probably, but have been holding off on that until taking a deeper view of how we specify dev envs.

[the remark this answers is not in the document]
Comment (alangenfeld) [reply .buildkite/dagster-buildkite/dagster_buildkite/steps/dagster.py:110]: What % of the total time is this `pip install` post uv? That would be the main motivator in my mind.

**Review by alangenfeld — APPROVED** [review]
sick

**Pull request merged**